README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

An example of PageView with page index animation in Flutter. There is way to add images from assets and short description as well. Support for carousel animation too.
Package support for local and network images both.

## Features

Support local images added in assets.
Support for network images.
Loading progress anination for high resolution images.
Ability to modify page index color, font, style etc.

## Getting started

Create a flutter project and in your pubspec.yaml add below dependency

```dependencies:
 zz_animate_pkg: ^0.0.6
```

## Usage

First create a List of map<String, String> as below

#### Note: Images are added in 'assets/images/' folder under root directory. Don't forget to add in assets of your pubspec.yaml file.

```
 List<Map<String, String>> imagesModel = [{
    'message': 'Ipsum is simply dummy text',
    'image': '1.jpg',
  },
  {
    'message':
        'There are many variations of passages of Lorem Ipsum.',
    'image': '2.jpg',
  },
  {
    'message':
        'It is a long established fact that a reader will be distracted.',
    'image': '3.jpg',
  },........]
```

For network images, create a list of map<String, String> as below

```
const List<Map<String, String>> networkImagesModel = [
{
  'message': 'Ipsum is simply dummy text of the printing',
  'image': 'https://images.pexels.com/photos/624015/pexels-photo-624015.jpeg',
},
{
  'message':
      'There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form',
  'image':
      'https://images.pexels.com/photos/3408744/pexels-photo-3408744.jpeg',
},....
```

Inside you class, add below code for Local images

```
  void main() {
   runApp(const MyApp());
 }

class MyApp extends StatelessWidget {
   const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
   @override
   Widget build(BuildContext context) {
   return MaterialApp(
     title: 'Pageview Demo',
      theme: ThemeData(
         primarySwatch: Colors.blue,
     ),
     home: PageViewScreen(
        imagesArray: imagesModel,
      ),
    );
  }
}
```

## Screenshot

<img width="424" alt="Screenshot 2021-12-21 at 11 46 01" src="https://user-images.githubusercontent.com/2304583/146917392-60b70966-8d6b-4246-a7e4-79f95ea05592.png">

For adding viewPort support

```
  void main() {
   runApp(const MyApp());
 }

class MyApp extends StatelessWidget {
   const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
   @override
   Widget build(BuildContext context) {
   return MaterialApp(
     title: 'Pageview Demo',
      theme: ThemeData(
         primarySwatch: Colors.blue,
     ),
     home: PageViewScreen(
        imagesArray: imagesModel,
        isViewFraction: true,
      ),
    );
  }
}
```

## Screenshot

<img width="428" alt="Screenshot 2021-12-21 at 11 43 46" src="https://user-images.githubusercontent.com/2304583/146916991-8bf747cd-a4de-4f14-86fa-07704f25eb75.png">

For network image support

```
  void main() {
   runApp(const MyApp());
 }

class MyApp extends StatelessWidget {
   const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
   @override
   Widget build(BuildContext context) {
   return MaterialApp(
     title: 'Pageview Demo',
      theme: ThemeData(
         primarySwatch: Colors.blue,
     ),
     home: PageViewScreen(
        imagesArray: networkImagesModel,
        isNetowrkImage: true,
      ),
    );
  }
}
```

## Additional information

HomeScreen parameters

```
isNetowrkImage: Bool Type: - By deafult 'false'. pass as 'true' value if network image support requires.
isViewFraction: Bool Type: - By deafult 'false'. Make 'true' if carusal view support requires.
cardCornerRadius: Double Type: - Default value is 8.0.
progressColor: Color Type: - Default value is 'red'. used as progress/animation color.
textStyle: TextStyle Type: - Text style of image's message.
viewportFraction: Double Type: - viewportFraction value of PageController. Default value is 0.8

```


CHANGELOG.md
## 0.0.6

Minor update and fixes.


test/zz_animate_pkg_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:zz_animate_pkg/zz_animate_pkg.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/zz_animate_pkg.dart
library zz_animate_pkg;

import 'package:flutter/material.dart';
import 'package:zz_animate_pkg/widget/footer_container.dart';
import 'package:zz_animate_pkg/widget/page_widget.dart';

class PageViewScreen extends StatefulWidget {
  final List<Map<String, String>> imagesArray;
  final pageController = PageController();
  final bool isViewFraction;
  final double cardCornerRadius;
  final double viewportFraction;
  final TextStyle textStyle;
  final bool isNetowrkImage;
  final Color progressColor;

  PageViewScreen({
    Key? key,
    required this.imagesArray,
    this.isViewFraction = false,
    this.cardCornerRadius = 0.0,
    this.viewportFraction = 0.8,
    this.textStyle = const TextStyle(
      color: Colors.white,
      fontSize: 14,
    ),
    this.isNetowrkImage = false,
    this.progressColor = Colors.red,
  }) : super(key: key);
  @override
  _PageViewScreenState createState() => _PageViewScreenState();
}

class _PageViewScreenState extends State<PageViewScreen> {
  late PageController pageController;
  int selectedIndex = 0;

  @override
  initState() {
    super.initState();
    pageController = widget.isViewFraction
        ? PageController(
            initialPage: selectedIndex,
            keepPage: true,
            viewportFraction: widget.viewportFraction,
          )
        : PageController();
  }

  @override
  dispose() {
    pageController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white.withAlpha(230),
      body: SafeArea(
        bottom: false,
        child: Column(
          children: [
            Expanded(
              child: PageViewWidget(
                imageArray: widget.imagesArray,
                scrollDirection: Axis.horizontal,
                pageSnapping: false,
                onPageChange: (val) => {onPageChange(val)},
                pageController: pageController,
                borderColor: Colors.grey,
                borderRadius: 10.0,
                borderWidth: 1.0,
                isViewPortVisible: widget.isViewFraction,
                textStyle: widget.textStyle,
                isNetworkImage: widget.isNetowrkImage,
                progressColor: widget.progressColor,
              ),
            ),
            FooterContainer(
              totalItems: widget.imagesArray.length,
              selectedIndex: selectedIndex,
            ),
          ],
        ),
      ),
    );
  }

  void onPageChange(int page) {
    setState(() {
      selectedIndex = page;
    });
  }
}


lib/widget/page_widget.dart
import 'package:flutter/material.dart';
import 'package:zz_animate_pkg/widget/image_info_widget.dart';

class PageViewWidget extends StatelessWidget {
  const PageViewWidget({
    Key? key,
    required this.imageArray,
    this.scrollDirection = Axis.horizontal,
    this.pageSnapping = true,
    required this.pageController,
    required this.onPageChange,
    required this.borderRadius,
    required this.borderWidth,
    required this.borderColor,
    required this.isViewPortVisible,
    required this.textStyle,
    required this.isNetworkImage,
    required this.progressColor,
  }) : super(key: key);

  final List<Map<String, String>> imageArray;
  final Axis scrollDirection;
  final bool pageSnapping;
  final Function onPageChange;
  final PageController pageController;
  final double borderRadius;
  final double borderWidth;
  final Color borderColor;
  final bool isViewPortVisible;
  final TextStyle textStyle;
  final bool isNetworkImage;
  final Color progressColor;

  @override
  Widget build(BuildContext context) {
    return PageView.builder(
      controller: pageController,
      scrollDirection: scrollDirection,
      onPageChanged: (page) {
        onPageChange(page);
      },
      itemBuilder: (cntx, index) {
        return Container(
          margin: const EdgeInsets.all(10),
          height: double.infinity,
          decoration: BoxDecoration(
              color: Colors.white,
              borderRadius: BorderRadius.circular(borderRadius + 2),
              border: Border.all(color: borderColor, width: borderWidth)),
          child: ClipRRect(
            borderRadius: BorderRadius.circular(borderRadius),
            child: Stack(
              fit: StackFit.expand,
              children: [
                Center(
                  child: CircularProgressIndicator(
                    color: progressColor,
                    strokeWidth: 2.0,
                  ),
                ),
                isNetworkImage
                    ? Image.network(
                        (imageArray.map((e) => e['image']!)).toList()[index],
                        fit: BoxFit.fill,
                      )
                    : Image.asset(
                        'assets/images/${(imageArray.map((e) => e['image']!)).toList()[index]}',
                        fit: BoxFit.fill,
                      ),
                ImageInfoWidget(
                  message:
                      (imageArray.map((e) => e['message']!)).toList()[index],
                  textStyle: textStyle,
                ),
              ],
            ),
          ),
        );
      },
      itemCount: imageArray.length,
    );
  }

  builder(int index) {
    return AnimatedBuilder(
      animation: pageController,
      builder: (context, child) {
        double value = 1.0;
        if (pageController.position.haveDimensions) {
          value = pageController.page! - index;
          value = (1 - (value.abs() * .5)).clamp(0.0, 1.0);
        } else {
          value = 0.6;
        }

        return Center(
          child: SizedBox(
            height: Curves.easeOut.transform(value) * 400,
            width: Curves.easeOut.transform(value) * 350,
            child: child,
          ),
        );
      },
      child: Container(
        margin: const EdgeInsets.all(8.0),
        color: index % 2 == 0 ? Colors.blue : Colors.red,
      ),
    );
  }
}


lib/widget/footer_container.dart
import 'package:flutter/material.dart';

class FooterContainer extends StatelessWidget {
  const FooterContainer({
    Key? key,
    required this.totalItems,
    required this.selectedIndex,
    this.defaultColor = Colors.grey,
    this.selectedColor = Colors.red,
  }) : super(key: key);

  final int totalItems;
  final int selectedIndex;
  final Color defaultColor;
  final Color selectedColor;
  @override
  Widget build(BuildContext context) {
    // ignore: dead_code
    return Container(
      alignment: Alignment.center,
      margin: const EdgeInsets.symmetric(horizontal: 20),
      width: double.infinity,
      height: 50,
      child: ListView.builder(
        shrinkWrap: true,
        scrollDirection: Axis.horizontal,
        itemBuilder: (cntx, index) {
          return Row(
            children: [
              AnimatedContainer(
                curve: Curves.fastOutSlowIn,
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(15 / 2),
                  color: index == selectedIndex ? selectedColor : defaultColor,
                ),
                duration: const Duration(milliseconds: 250),
                height: 15,
                width: index == selectedIndex ? 30 : 15,
              ),
              const SizedBox(
                width: 5,
              )
            ],
          );
        },
        itemCount: totalItems,
      ),
    );
  }
}


example/lib/model/image_model.dart
const List<Map<String, String>> networkImagesModel = [
  {
    'message': 'Ipsum is simply dummy text of the printing',
    'image': 'https://images.pexels.com/photos/624015/pexels-photo-624015.jpeg',
  },
  {
    'message':
        'There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form',
    'image':
        'https://images.pexels.com/photos/3408744/pexels-photo-3408744.jpeg',
  },
  {
    'message':
        'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout',
    'image':
        'https://images.pexels.com/photos/1509582/pexels-photo-1509582.jpeg',
  },
  {
    'message':
        'Translations: Can you help translate this site into a foreign language ? Please email us with details if you can help.',
    'image':
        'https://images.pexels.com/photos/3244513/pexels-photo-3244513.jpeg',
  },
  {
    'message':
        'The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested.',
    'image': 'https://images.pexels.com/photos/572897/pexels-photo-572897.jpeg',
  },
  {
    'message':
        'A dictionary of over 100  words, combined with a handful of model  structures',
    'image':
        'https://images.pexels.com/photos/1761279/pexels-photo-1761279.jpeg',
  },
  {
    'message': 'Ipsum is simply dummy text of the printing',
    'image':
        'https://images.pexels.com/photos/3900437/pexels-photo-3900437.jpeg',
  },
  {
    'message': 'Ipsum is simply dummy text of the printing',
    'image':
        'https://images.pexels.com/photos/1770809/pexels-photo-1770809.jpeg',
  },
  {
    'message':
        'The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested.',
    'image':
        'https://images.pexels.com/photos/3876407/pexels-photo-3876407.jpeg',
  },
];

// Network image model

const List<Map<String, String>> imagesModel = [
  {
    'message': 'Ipsum is simply dummy text of the printing',
    'image': '1.jpg',
  },
  {
    'message':
        'There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form',
    'image': '2.jpg',
  },
  {
    'message':
        'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout',
    'image': '3.jpg',
  },
  {
    'message':
        'Translations: Can you help translate this site into a foreign language ? Please email us with details if you can help.',
    'image': '4.jpg',
  },
  {
    'message':
        'The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested.',
    'image': '5.jpg',
  },
  {
    'message':
        'A dictionary of over 100  words, combined with a handful of model  structures',
    'image': '6.jpg',
  },
  {
    'message': 'Ipsum is simply dummy text of the printing',
    'image': '7.jpg',
  },
  {
    'message': 'Ipsum is simply dummy text of the printing',
    'image': '8.jpg',
  },
  {
    'message':
        'The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested.',
    'image': '9.jpg',
  },
  {
    'message':
        'Sections 1.10.32 and 1.10.33 from "de Finibus Bonorum et Malorum" by Cicero are also reproduced in their exact original form',
    'image': '10.jpg',
  },
  {
    'message':
        'It uses a dictionary of over 200 Latin words, combined with a handful of model sentence structures',
    'image': '11.jpg',
  },
  {
    'message':
        'Accompanied by English versions from the 1914 translation by H. Rackham',
    'image': '12.jpg',
  },
  {
    'message': 'Ipsum is simply dummy text of the printing',
    'image': '13.jpg',
  },
  {
    'message':
        'From "de Finibus Bonorum et Malorum" by Cicero are also reproduced in their exact original form',
    'image': '14.jpg',
  },
  {
    'message':
        ' typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s',
    'image': '15.jpg',
  },
];
